modules/standard/BigInteger.chpl

```diff
@@ -3535,508 +3649,421 @@ module BigInteger {
       }
     }
   }
 
-  proc bigint.mul(const ref a: bigint, const ref b: bigint) {
-    if _local {
-      mpz_mul(this.mpz, a.mpz, b.mpz);
+  @deprecated(notes="bigint.sub method is deprecated - please use the standalone function :proc:`~BigInteger.sub`")
+  proc bigint.sub(a: uint, const ref b: bigint) {
+    BigInteger.sub(this, a, b);
+  }
 
-    } else if this.localeId == chpl_nodeID &&
+  proc mul(ref result: bigint, const ref a: bigint, const ref b: bigint) {
+    if _local {
+      mpz_mul(result.mpz, a.mpz, b.mpz);
+    } else if result.localeId == chpl_nodeID &&
               a.localeId    == chpl_nodeID &&
               b.localeId    == chpl_nodeID {
-      mpz_mul(this.mpz, a.mpz, b.mpz);
-
+      mpz_mul(result.mpz, a.mpz, b.mpz);
     } else {
-      const thisLoc = chpl_buildLocaleID(this.localeId, c_sublocid_any);
-
-      on __primitive("chpl_on_locale_num", thisLoc) {
-        const a_ = a;
-        const b_ = b;
-
-        mpz_mul(this.mpz, a_.mpz, b_.mpz);
+      const resultLoc = chpl_buildLocaleID(result.localeId, c_sublocid_any);
+      on __primitive("chpl_on_locale_num", resultLoc) {
+        const a_ = a.localize();
+        const b_ = b.localize();
+        mpz_mul(result.mpz, a_.mpz, b_.mpz);
       }
     }
   }
 
-  proc bigint.mul(const ref a: bigint, b: int) {
+  @deprecated(notes="bigint.mul method is deprecated - please use the standalone function :proc:`~BigInteger.mul`")
+  proc bigint.mul(const ref a: bigint, const ref b: bigint) {
+    BigInteger.mul(this, a, b);
+  }
+
+  proc mul(ref result: bigint, const ref a: bigint, b: int) {
     const b_ = b.safeCast(c_long);
 
     if _local {
-      mpz_mul_si(this.mpz, a.mpz, b_);
-
-    } else if this.localeId == chpl_nodeID && a.localeId == chpl_nodeID {
-      mpz_mul_si(this.mpz, a.mpz, b_);
-
+      mpz_mul_si(result.mpz, a.mpz, b_);
+    } else if result.localeId == chpl_nodeID && a.localeId == chpl_nodeID {
+      mpz_mul_si(result.mpz, a.mpz, b_);
     } else {
-      const thisLoc = chpl_buildLocaleID(this.localeId, c_sublocid_any);
-
-      on __primitive("chpl_on_locale_num", thisLoc) {
-        const a_ = a;
-
-        mpz_mul_si(this.mpz, a_.mpz, b_);
+      const resultLoc = chpl_buildLocaleID(result.localeId, c_sublocid_any);
+      on __primitive("chpl_on_locale_num", resultLoc) {
+        const a_ = a.localize();
+        mpz_mul_si(result.mpz, a_.mpz, b_);
       }
     }
   }
 
-  proc bigint.mul(const ref a: bigint, b: uint) {
-    const b_ = b.safeCast(c_ulong);
-
-    if _local {
-      mpz_mul_ui(this.mpz, a.mpz, b_);
+  @deprecated(notes="bigint.mul method is deprecated - please use the standalone function :proc:`~BigInteger.mul`")
+  proc bigint.mul(const ref a: bigint, b: int) {
+    BigInteger.mul(this, a, b);
+  }
 
-    } else if this.localeId == chpl_nodeID && a.localeId == chpl_nodeID {
-      mpz_mul_ui(this.mpz, a.mpz, b_);
+  proc mul(ref result: bigint, const ref a: bigint, b: uint) {
+    const b_ = b.safeCast(c_long);
```

Comment:
@jabraham17 : Shouldn't this be doing a safe cast to `c_ulong` rather than `c_long`?

Reply:
Yes it should. I'll fix that right now

Reply:
This will be fixed by https://github.com/chapel-lang/chapel/pull/26295